{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# hyungtai sim, 24.04.18.\n",
    "\n",
    "vlib = c(\"tidyverse\", \"mashr\", \"ashr\", \"ggpubr\", \"data.table\", \"patchwork\", \"tidyseurat\",\n",
    "         \"future.apply\", \"arrow\", 'pheatmap', \"Seurat\", \"hdWGCNA\", \"enrichR\", \"SeuratData\",\n",
    "         \"Azimuth\", \"data.table\", \"presto\", \"Matrix\")\n",
    "lapply(vlib, require, character.only = TRUE, quietly = TRUE) |> suppressMessages()\n",
    "\n",
    "base.dir=\"/path/to/data\"\n",
    "chr_wd = paste0(base.dir, \"202404-sceQTLv7/\")\n",
    "setwd(chr_wd)\n",
    "\n",
    "saveRDS.gz <- function(object,file,threads=parallel::detectCores()) {\n",
    "  con <- pipe(paste0(\"pigz -p\",threads,\" > \",file),\"wb\")\n",
    "  saveRDS(object, file = con)\n",
    "  close(con)\n",
    "}\n",
    "\n",
    "readRDS.gz <- function(file,threads=parallel::detectCores()) {\n",
    "  con <- pipe(paste0(\"pigz -d -c -p\",threads,\" \",file))\n",
    "  object <- readRDS(file = con)\n",
    "  close(con)\n",
    "  return(object)\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## aggregation functions ####\n",
    "\n",
    "aggregate_counts = function(count_mat, annot, method = 'sum', verbose = TRUE) {\n",
    "  annot = annot %>% mutate(group = factor(group))\n",
    "  cells = intersect(colnames(count_mat), annot$cell)\n",
    "  count_mat = count_mat[,cells]\n",
    "  annot = annot %>% filter(cell %in% cells)\n",
    "  cell_dict = annot %>% {setNames(.$group, .$cell)} %>% droplevels\n",
    "  cell_dict = cell_dict[cells]\n",
    "  if (verbose) {\n",
    "    print(table(cell_dict))\n",
    "  }\n",
    "  if (length(levels(cell_dict)) == 1) {\n",
    "    count_mat_clust = count_mat %>% rowSums() %>% as.matrix %>% magrittr::set_colnames(levels(cell_dict))\n",
    "    mean_mat_clust = count_mat_clust / as.numeric(table(cell_dict))\n",
    "  } else {\n",
    "    M = model.matrix(~ 0 + cell_dict) %>% magrittr::set_colnames(levels(cell_dict)) %>% Matrix(., sparse = TRUE)\n",
    "    mtx_diag_cellcount = diag(1/table(cell_dict)) %>% Matrix(., sparse = TRUE)\n",
    "    count_mat_clust = count_mat %*% M %>% Matrix(., sparse = TRUE)\n",
    "    mean_mat_clust = count_mat_clust %*% mtx_diag_cellcount %>% Matrix(., sparse = TRUE)\n",
    "  }\n",
    "  if (method == 'sum') {\n",
    "    return(count_mat_clust)\n",
    "  } else if (method == 'mean') {\n",
    "    return(mean_mat_clust)\n",
    "  } else {\n",
    "    cat(\"set the method in sum or mean. \\n\")\n",
    "    stop()\n",
    "  }\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "pull_barcode = function(metadata){\n",
    "  results = metadata %>% \n",
    "  dplyr::select(CellID, sample_time) %>%\n",
    "  dplyr::rename(cell =  CellID, group = sample_time) %>% \n",
    "  group_by(group) %>% add_count() %>% filter(n > 3) %>%\n",
    "  select(-n)\n",
    "  return(results)\n",
    "}\n",
    "\n",
    "create_pseudobulk = function(df_barcode, dgem_full){\n",
    "  int_len = df_barcode %>% distinct(group) %>% pull(group) %>% length()\n",
    "  dgem_filtered= dgem_full[,df_barcode$cell]\n",
    "  pseudobulk_filtered_count = aggregate_counts(dgem_filtered > 0, df_barcode)\n",
    "  # filter 2.\n",
    "  ## we will count the number of cells with counts >= 1 each gene, from each sample-cluster\n",
    "  ## we collect the expressed cell count and filter >=3, which indicates 3 or more cells express specific genes per sample-cluster combination\n",
    "  pseudobulk_results = aggregate_counts(dgem_filtered, df_barcode, method = \"mean\")\n",
    "  #pseudobulk_results[pseudobulk_filtered_count < 3] <- 0\n",
    "  pseudobulk_filtered_genes = rowSums(pseudobulk_results > 0) %>% as.data.frame() %>%\n",
    "    rownames_to_column(\"gene_name\") %>% \n",
    "    filter(`.` >= 1.0*int_len) %>% pull(gene_name)\n",
    "  colnames(pseudobulk_results) = colnames(pseudobulk_filtered_count)\n",
    "  pseudobulk_results = pseudobulk_results[pseudobulk_filtered_genes,]\n",
    "  return(pseudobulk_results)\n",
    "}\n",
    "\n",
    "plot_logexpression = function(pseudobulk_results) {\n",
    "  p1 = pseudobulk_results %>% as.matrix() %>% as.data.frame() %>%\n",
    "    rownames_to_column('gene_name') %>%\n",
    "    pivot_longer(cols = c(2:ncol(.))) %>%\n",
    "    group_by(name) %>%\n",
    "    filter(value > 0) %>%\n",
    "    ggplot(aes(x = name, y = value)) + \n",
    "    geom_boxplot() + \n",
    "    theme_pubr(x.text.angle = 90) +\n",
    "    scale_y_log10()\n",
    "  return(p1)\n",
    "}\n",
    "\n",
    "plot_pca = function(pseudobulk_results) {\n",
    "  results_pca = vector(mode = \"list\", length = 3L)\n",
    "  pca_obj = pseudobulk_results %>% prcomp()\n",
    "  pca_obj_summary = pca_obj %>% \n",
    "    summary() %>% \n",
    "    .$importance %>% \n",
    "    t() %>% \n",
    "    as.data.frame()\n",
    "  \n",
    "  p1 = pca_obj$rotation %>% \n",
    "    as.data.frame() %>%\n",
    "    ggplot(aes(x = PC1, y = PC2)) + \n",
    "    geom_point() +\n",
    "    theme_pubr() +\n",
    "    xlab(paste0(\"PC1 (\",\n",
    "                format(pca_obj_summary$`Proportion of Variance`[1]*100, digit=3),\n",
    "                \" % of Variance)\")) +\n",
    "    ylab(paste0(\"PC2 (\", \n",
    "                format(pca_obj_summary$`Proportion of Variance`[2]*100, digit=3),\n",
    "                \" % of Variance)\"))\n",
    "  \n",
    "  results_pca$pca = pca_obj\n",
    "  results_pca$var_prop = pca_obj_summary\n",
    "  results_pca$plot = p1\n",
    "  print(p1)\n",
    "  return(results_pca)\n",
    "}\n",
    "\n",
    "create_genotype = function(chr_path_si, chr_path_fam, df_sample_name, n_max = 73, chr_prefix, global_prefix){\n",
    "  temp_prefix = \"/home/rstudio/.local/temp/\" # change if needed\n",
    "  list_res = vector(mode = \"list\", length = 2L)\n",
    "  names(list_res) = c(\"intersect\", \"genotype_prefix\")\n",
    "  df_si = read_delim(chr_path_si) %>%\n",
    "    filter(IID %in% df_sample_name$sample) %>%\n",
    "    select(IID, response, age, sex, histology)\n",
    "  df_geno_all = read_delim(chr_path_fam, col_names = F)\n",
    "  colnames(df_geno_all) = c(\"FID\", \"IID\", \"V3\", \"V4\", \"V5\", \"V6\")\n",
    "  intersect = left_join(df_si,df_geno_all, by = c(\"IID\")) %>%\n",
    "    mutate(sex = ifelse(sex == \"M\", \"1\", ifelse(sex == \"F\", \"2\", \"0\"))) \n",
    "  list_res$intersect = intersect\n",
    "  ### check module if N = 73 (almost all) ####\n",
    "  if (nrow(df_si) == n_max){\n",
    "    # we don't need any precessing, since all samples are listed.\n",
    "    #    subset_ids = intersect %>% \n",
    "    #      mutate(before_FID = IID, before_IID = IID) %>% \n",
    "    #      select(before_FID, before_IID) %>%\n",
    "    #      distinct(before_FID, before_IID) %>% \n",
    "    #      write_delim(paste0(temp_prefix, \"temp_plink.subset_ids.txt\"), col_names = FALSE, delim = '\\t')\n",
    "    #    update_ids = intersect %>% \n",
    "    #      mutate(before_FID = 0, before_IID = IID, after_FID = IID, after_IID = IID) %>% \n",
    "    #      select(before_FID, before_IID, after_FID, after_IID) %>%\n",
    "    #      distinct(before_FID, before_IID, .keep_all = TRUE) %>%\n",
    "    #      write_delim(paste0(temp_prefix, \"temp_plink.update_ids.txt\"), col_names = FALSE, delim = '\\t')\n",
    "    #    update_sex = intersect %>% \n",
    "    #      mutate(FID = IID, IID = IID) %>% \n",
    "    #      select(FID, IID, sex) %>%\n",
    "    #      distinct(FID, IID, .keep_all = TRUE) %>%\n",
    "    #      write_delim(paste0(temp_prefix,\"temp_plink.update_sex.txt\"), col_names = FALSE, delim = '\\t')\n",
    "    #    system(paste0(\"analysis/00_02_01_plink_script.sh \", chr_prefix))\n",
    "    list_res$genotype_prefix = paste0(global_prefix, \"geno/all_plink_maf01_HWEe5_geno03_mind03\")\n",
    "  } else {\n",
    "    subset_ids = intersect %>% \n",
    "      mutate(before_FID = IID, before_IID = IID) %>% \n",
    "      select(before_FID, before_IID) %>%\n",
    "      distinct(before_FID, before_IID) %>% \n",
    "      write_delim(paste0(temp_prefix, \"temp_plink.subset_ids.txt\"), col_names = FALSE, delim = '\\t')\n",
    "    update_ids = intersect %>% \n",
    "      mutate(before_FID = 0, before_IID = IID, after_FID = IID, after_IID = IID) %>% \n",
    "      select(before_FID, before_IID, after_FID, after_IID) %>%\n",
    "      distinct(before_FID, before_IID, .keep_all = TRUE) %>%\n",
    "      write_delim(paste0(temp_prefix, \"temp_plink.update_ids.txt\"), col_names = FALSE, delim = '\\t')\n",
    "    update_sex = intersect %>% \n",
    "      mutate(FID = IID, IID = IID) %>% \n",
    "      select(FID, IID, sex) %>%\n",
    "      distinct(FID, IID, .keep_all = TRUE) %>%\n",
    "      write_delim(paste0(temp_prefix,\"temp_plink.update_sex.txt\"), col_names = FALSE, delim = '\\t')\n",
    "    system(paste0(\"analysis/00_02_01_plink_script.sh \", chr_prefix))\n",
    "    list_res$genotype_prefix = paste0(global_prefix, \"plink_subset/\", chr_prefix, \"_maf01_HWEe5_geno03_mind03\")\n",
    "  }\n",
    "  return(list_res)\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# spare matrix reading and filtering ####\n",
    "\n",
    "## read sparse matrix in SCTransformed raw count. ####\n",
    "\n",
    "dgem = readRDS.gz(\"01_calling_eQTL/01_00_pseudobulk_pipeline/all_dgcmatrix_for_sceQTLv6a.rds\")\n",
    "\n",
    "## 1% or more expressed genes in cohort; generates character vector with genes.  ####\n",
    "count_filtered_genes = rowSums(dgem > 0) %>%\n",
    "  as.data.frame() %>%\n",
    "  rownames_to_column(\"gene_name\") %>%\n",
    "  filter(`.` > 0.01 * dim(dgem)[2]) %>%\n",
    "  pull(gene_name)\n",
    "dgem = dgem[count_filtered_genes, ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# 2. iter.tools ####\n",
    "\n",
    "## 2.0 일반적으로 Cluster level로 pipeline을 짜야 함. ####\n",
    "### genotype :\n",
    "### phenotype : cell id도 있으면 좋음 \n",
    "### covariates :\n",
    "#### genotype PCA, other MD : id만 필요\n",
    "#### phenotype PCA : phenotype 전체가 필요함\n",
    "\n",
    "## as a pipeline : phenotype 에서 나머지 만들기\n",
    "\n",
    "global_prefix = \"01_calling_eQTL/01_00_pseudobulk_pipeline/\"\n",
    "gtf = read_delim(\"gencode.v43.basic.tss_bed.txt\")\n",
    "\n",
    "chr_cluster_definition = \"anno_l1\"\n",
    "chr_time_definition = \"1st\" # base for baseline, or 1st for treatment\n",
    "chr_sampleid = \"sample_time\"\n",
    "chr_path_sample_information = \"00_sample_information_v8.txt\"\n",
    "# note: this is same as 220426_maf_0.1_hwe_1e-5_genomind_0.03_updateIID\n",
    "chr_original_fam = \"/path/to/data/202404-sceQTLv7/01_calling_eQTL/01_00_pseudobulk_pipeline/original/230628_chrALL_sample_ALLIO.fam\"\n",
    "\n",
    "\n",
    "## 2.1. define cluster, barcode df and sample list  ####\n",
    "\n",
    "# @ 240123 modified. join-all and select\n",
    "\n",
    "c_clusters = df_full_md_annotated %>% \n",
    "  group_by(!!! rlang::syms(c(chr_sampleid, chr_cluster_definition, \"time\"))) %>%\n",
    "  summarise(n = n()) %>%\n",
    "  filter(time == chr_time_definition, n >= 3) %>% ungroup() %>% select(-n) %>%\n",
    "  group_by(get(chr_cluster_definition)) %>% add_count() %>% filter(n >= 0.85 * 73) %>%\n",
    "  distinct(get(chr_cluster_definition)) %>% pull(.)\n",
    "\n",
    "\n",
    "for (idx in c(1:length(c_clusters))){\n",
    "\n",
    "  c_each_cluster = c_clusters[idx]\n",
    "  list_path_res = vector(mode = \"list\", length = length(c_each_cluster))\n",
    "  names(list_path_res) = c_each_cluster\n",
    "  \n",
    "  # check if needed\n",
    "  chr_prefix = paste(chr_cluster_definition, chr_time_definition, c_each_cluster, sep = \"_\")\n",
    "  chr_exprs_path = paste0(global_prefix,\"exprs_matrix/\", chr_prefix, \".bed\")\n",
    "  chr_raw_exprs_path = paste0(global_prefix,\"exprs_matrix_raw/\", chr_prefix, \".bed\")\n",
    "  chr_covs_path = paste0(global_prefix,\"covariates/\", chr_prefix, \".txt\")\n",
    "  chr_interaction_path = paste0(global_prefix,\"interactions/\", chr_prefix, \".txt\")\n",
    "  chr_qtlinput_path = paste0(global_prefix, \"tensorqtl_input_l2/\", chr_prefix, \".txt\")\n",
    "  \n",
    "  df_groupinfo = df_full_md_annotated %>% \n",
    "    filter(sample %in% chr_all_samples) %>%\n",
    "    #filter(is.na(CHIP_BEFORE_BINARY) == FALSE) %>% # as we don't need any interaction pipeline for NOW.\n",
    "    filter(time %in% chr_time_definition, get(chr_cluster_definition) %in% c_each_cluster) %>%\n",
    "    pull_barcode()\n",
    "  \n",
    "  df_sample_name = df_groupinfo %>% \n",
    "    separate(group, into = c(\"sample\", \"time\"), sep = \"-\", remove = F) %>% ungroup() %>% distinct(sample)\n",
    "  \n",
    "  ## 2.2. sample sex annotation for genotype QC ####\n",
    "  \n",
    "  list_geno_res = create_genotype(\n",
    "    chr_path_si = chr_path_sample_information,\n",
    "    chr_path_fam = chr_original_fam, \n",
    "    df_sample_name = df_sample_name,\n",
    "    n_max = 73\n",
    "    chr_prefix = chr_prefix,\n",
    "    global_prefix = global_prefix\n",
    "  )\n",
    "  \n",
    "  ## 2.3. expression module ####\n",
    "  \n",
    "  pseudobulk_results = create_pseudobulk(df_groupinfo, dgem)\n",
    "  pca_pseudobulk = plot_pca(pseudobulk_results)\n",
    "  \n",
    "  raw_exprs_df = pseudobulk_results %>% \n",
    "    as.matrix() %>% as.data.frame() %>% rownames_to_column(\"gene_id\") %>%\n",
    "    pivot_longer(cols = c(2:ncol(.))) %>%\n",
    "    group_by(gene_id)\n",
    "  \n",
    "  raw_exprs_df %>% separate(name, into = c(\"name\", \"time\"), sep = \"-\") %>% select(-time) %>%\n",
    "    pivot_wider(names_from = name, values_from = value) %>%\n",
    "    right_join(gtf, .) %>%\n",
    "    select(1:4, df_sample_name$sample) %>%\n",
    "    dplyr::rename(phenotype_id = gene_id) %>%\n",
    "    distinct(`#chr`, start, end, .keep_all=TRUE) %>% add_count(phenotype_id) %>%\n",
    "    filter(is.na(start) == FALSE, n ==1) %>% select(-n) %>%\n",
    "    write_delim(chr_raw_exprs_path, delim = \"\\t\", quote = \"none\")\n",
    "  \n",
    "  qnorm_df = raw_exprs_df %>%\n",
    "    mutate(norm = qnorm((rank(value,na.last=\"keep\")-0.5)/sum(!is.na(value)))) %>%\n",
    "    select(-value) %>%\n",
    "    separate(name, into = c(\"name\", \"time\"), sep = \"-\") %>% select(-time) %>%\n",
    "    pivot_wider(names_from = name, values_from = norm)\n",
    "  \n",
    "  right_join(gtf, qnorm_df) %>% select(1:4, df_sample_name$sample) %>%\n",
    "    dplyr::rename(phenotype_id = gene_id) %>%\n",
    "    distinct(`#chr`, start, end, .keep_all=TRUE) %>% add_count(phenotype_id) %>%\n",
    "    filter(is.na(start) == FALSE, n ==1) %>% select(-n) %>%\n",
    "    write_delim(chr_exprs_path, delim = \"\\t\", quote = \"none\")\n",
    "  \n",
    "  # note : use miniconda to install tabix, bgzip and plink if needed\n",
    "  system(paste0('~/miniconda3/bin/bgzip -f ', chr_exprs_path))\n",
    "  system(paste0('~/miniconda3/bin/tabix -p bed -f ', chr_exprs_path, \".gz\"))\n",
    "\n",
    "  ## 2.4. Covariates module ####\n",
    "  \n",
    "  df_cov_base = fread(paste0(list_geno_res$genotype_prefix, \".fam\")) %>% select(V2, V5) %>% \n",
    "    filter(V2 %in% df_sample_name$sample) %>% \n",
    "    mutate(V2 = factor(V2, levels = df_sample_name$sample)) %>%\n",
    "    arrange(V2)%>%\n",
    "    dplyr::rename(sex = V5, IID = V2)\n",
    "  \n",
    "  df_ePCA = pca_pseudobulk$pca$rotation %>% as.data.frame() %>% rownames_to_column(\"id_base\") %>%\n",
    "    separate(id_base, sep = \"-\", into = c(\"IID\", \"time\")) %>%\n",
    "    select(1,PC1:PC10) # 10 PCs\n",
    "  \n",
    "  df_gPCA =  fread(paste0(list_geno_res$genotype_prefix, \".eigenvec\")) %>%\n",
    "    select(1, V3:V5) %>% # 3 gPCs\n",
    "    dplyr::rename(IID = V1)\n",
    "  \n",
    "  df_si = read_delim(chr_path_sample_information) %>% \n",
    "    select(IID, response, histology, age)\n",
    "  \n",
    "  df_file =  df_full_md_annotated %>%  \n",
    "    distinct(sample_time, file, CHIP_BEFORE_BINARY) %>%\n",
    "    separate(sample_time, into = c(\"name\", \"time\"), sep = \"-\") %>% \n",
    "    filter(time == chr_time_definition, name %in% df_sample_name$sample) %>%\n",
    "    select(-time) %>%\n",
    "    dplyr::rename(IID = name) %>%\n",
    "    mutate(CHIP_BEFORE_BINARY = ifelse(CHIP_BEFORE_BINARY== TRUE, 1, 0))\n",
    "  \n",
    "  df_cov = left_join(df_cov_base, df_ePCA) %>%\n",
    "    left_join(., df_gPCA) %>%\n",
    "    left_join(., df_si) %>%\n",
    "    left_join(., df_file) %>%\n",
    "    mutate(sex = sex -1) \n",
    "  \n",
    "  df_cov %>% select(1:V5, age, file) %>%\n",
    "    t() %>%\n",
    "    write.table(chr_covs_path, quote = FALSE, col.names = F, sep = \"\\t\") # gPC 3, ePC 10, sex \n",
    "  \n",
    "  df_cov %>% select(1, response, histology) %>% t() %>%\n",
    "    write.table(chr_interaction_path, quote = FALSE, col.names = F, sep = \"\\t\")\n",
    "  # sex is not used for interaction term. cannot be stratified for eqtl.\n",
    "  # smoking is also not used for interaction term. cannot be stratified for eqtl.\n",
    "  \n",
    "  df_output_list = data.frame(\n",
    "    geno_prefix = list_geno_res$genotype_prefix,\n",
    "    pheno_path = paste0(chr_exprs_path, \".gz\"),\n",
    "    covs_path = chr_covs_path,\n",
    "    interaction_path = chr_interaction_path,\n",
    "    raw_pheno_path = chr_raw_exprs_path\n",
    "  ) %>% pivot_longer(cols = 1:5)\n",
    "\n",
    "\n",
    "  df_output_list %>% write_delim(chr_qtlinput_path, delim = \"\\t\")\n",
    "  \n",
    "}\n"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
